price-feeder/oracle/provider/binance.go

```diff
@@ -23,96 +26,185 @@ type (
-	path := fmt.Sprintf("%s%s?symbol=%s", p.baseURL, binanceTickerEndpoint, ticker)
-
-	resp, err := p.client.Get(path)
-	if err != nil {
-		return TickerPrice{}, fmt.Errorf("failed to make Binance request: %w", err)
+func (p *BinanceProvider) getTickerPrice(key string) (TickerPrice, error) {
+	ticker, ok := p.tickers[key]
+	if !ok {
+		return TickerPrice{}, fmt.Errorf("failed to get %s", key)
 	}
 
-	defer resp.Body.Close()
+	return ticker.toTickerPrice()
+}
 
-	bz, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		return TickerPrice{}, fmt.Errorf("failed to read Binance response body: %w", err)
+func (p *BinanceProvider) messageReceived(messageType int, bz []byte) {
+	if messageType != websocket.TextMessage {
+		return
 	}
 
-	var tickerResp BinanceTickerResponse
-	if err := json.Unmarshal(bz, &tickerResp); err != nil {
-		return TickerPrice{}, fmt.Errorf("failed to unmarshal Binance response body: %w", err)
+	var tickerRespWS BinanceTicker
+	if err := json.Unmarshal(bz, &tickerRespWS); err != nil {
+		// sometimes it returns other messages which are not ticker responses
+		p.logger.Err(err).Msg("Binance provider could not unmarshal")
+		return
 	}
 
-	if tickerResp.Code != 0 {
-		return TickerPrice{}, fmt.Errorf(
-			"received unexpected error from Binance response: %v (%d)",
-			tickerResp.Msg, tickerResp.Code,
-		)
+	if len(tickerRespWS.LastPrice) == 0 {
+		return
 	}
 
-	if !strings.EqualFold(tickerResp.Symbol, ticker) {
-		return TickerPrice{}, fmt.Errorf(
-			"received unexpected symbol from Binance response; expected: %s, got: %s",
-			ticker, tickerResp.Symbol,
-		)
+	p.setTickerPair(tickerRespWS)
+}
+
+func (p *BinanceProvider) setTickerPair(ticker BinanceTicker) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.tickers[ticker.Symbol] = ticker
+}
+
+func (ticker BinanceTicker) toTickerPrice() (TickerPrice, error) {
+	return newTickerPrice("Binance", ticker.Symbol, ticker.LastPrice, ticker.Volume)
+}
+
+// subscribeTickers subscribe to all currency pairs
+func (p *BinanceProvider) subscribeTickers(cps ...types.CurrencyPair) error {
+	params := make([]string, len(cps))
+
+	for i, cp := range cps {
+		params[i] = strings.ToLower(cp.Base + cp.Quote + "@ticker")
 	}
 
-	price, err := sdk.NewDecFromStr(tickerResp.LastPrice)
-	if err != nil {
-		return TickerPrice{}, fmt.Errorf("failed to parse Binance price (%s) for %s", tickerResp.LastPrice, ticker)
+	subsMsg := newBinanceSubscriptionMsg(params...)
+	return p.wsClient.WriteJSON(subsMsg)
+}
+
+func (p *BinanceProvider) handleReceivedTickers(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(defaultReadNewMessage):
+			// time after to avoid asking for prices too frequently
+			messageType, bz, err := p.wsClient.ReadMessage()
+			if err != nil {
+				// if some error occurs continue to try to read the next message
+				p.logger.Err(err).Msg("Binance provider could not read message")
```

Comment:
What type is `messageType`? Maybe we should log that too.

Reply:
```go
// The message types are defined in RFC 6455, section 11.8.
const (
	// TextMessage denotes a text data message. The text message payload is
	// interpreted as UTF-8 encoded text data.
	TextMessage = 1

	// BinaryMessage denotes a binary data message.
	BinaryMessage = 2

	// CloseMessage denotes a close control message. The optional message
	// payload contains a numeric code and text. Use the FormatCloseMessage
	// function to format a close message payload.
	CloseMessage = 8

	// PingMessage denotes a ping control message. The optional message payload
	// is UTF-8 encoded text.
	PingMessage = 9

	// PongMessage denotes a pong control message. The optional message payload
	// is UTF-8 encoded text.
	PongMessage = 10
)

```
Maybe, what do you think?